docs/proposals/distributed-query-execution.md

```diff
@@ -0,0 +1,197 @@
+These fragments are then **enqueued** into the scheduler’s internal queue system, ready to be picked up by queriers.
+
+![SchedulerQueue](/images/proposals/distributed-execution-scheduler-queue.png)
+
+
+#### Assigning query fragments to queriers
+
+The fragment include the plan and the remote fragment IDs it depends on. Queriers pull fragments from the scheduler. After child fragments are assigned to a particular querier, it’s parent fragment is updated with the location of the querier from which to read the results from.
+
+At execution time, a `Remote` operator acts as a placeholder within the querier's execution engine to **pull results from other queriers** running dependent fragments.
+
+![AssignedFragments](/images/proposals/distributed-execution-assigned-fragments.png)
+
+### Querier
+
+The **Querier** is enhanced to support both **executing** query fragments and **serving** results to other queriers. In the distributed query model, a querier may either evaluate a plan fragment or provide intermediate results to other queriers that depend on its output.
+
+To support this, the querier now implements two interfaces:
+* **QueryExecutor** – Executes assigned query plan root fragment.
+* **QueryServer** – Serves results to other queriers via streaming gRPC endpoints. (i.e., acts as a remote data source for `Remote` operators).
+
+#### QueryExecutor
+The QueryExecutor interface enables a querier to execute the root fragment of the distributed query plan.
+
+```
+type QueryExecutor interface {
+	ExecuteQuery(p LogicalPlan) (QueryResult, error)
+}
+```
+
+#### QueryServer
+The QueryServer interface enables a querier to expose query results to other queriers via a gRPC API. This is used when another querier is executing a fragment that depends on the results of this one (i.e., the child in the execution tree).
+
+```
+type QueryServer interface {
+	Enqueue(context.Context, p LogicalPlan) error
+    Series(SeriesRequest, querierpb.QueryServer_SeriesServer) error
+	Next(*executionpb.NextRequest, querierpb.QueryServer_NextServer) error
+}
+```
+
+* **Enqueue** - Schedules the provided logical plan fragment for execution on the querier. Enqueuing a plan on a quereir will block a concurrency until the query finishes, or times out if it exceeds the configured limit. It respects the querier’s concurrency limits and queue capacity.
+* **Series** - Streams the metadata (labels) for the series produced by the fragment, allowing parent queriers to understand the shape of incoming data.
+* **Next** - Streams the actual step vector results of the fragment, returning time-step-wise computed samples and histograms.
+
+These methods are implemented as streaming gRPC endpoints and used by `Remote` operators to pull data on-demand during execution.
+
+
+##### gRPC Service Definition
+
+```
+service QueryServer {
+  rpc Series(SeriesRequest) returns (stream OneSeries);
```

Comment:
This should stream a batch of series, not just one?

Reply:
I think we can let grpc handle the batching in the transport layer. Store-gateways use this approach today.

https://github.com/cortexproject/cortex/blob/c1f2190e41e6a1f10db23435c8bd9357e0f9726c/pkg/storegateway/storegatewaypb/gateway.proto#L16